src/utils/decorators.ts

```diff
@@ -0,0 +1,29 @@
+export function memoize() {
+  return function (
+    target: any,
+    propertyKey: string,
+    descriptor: PropertyDescriptor
+  ) {
+    if (!descriptor.get) {
+      throw new Error(`@memoize can only be used for property accessors`);
+    }
+    let originalGet = descriptor.get;
+    descriptor.get = memoizedAccessor(originalGet, propertyKey);
+  };
+}
+
+function memoizedAccessor(
+  originalGet: Function,
+  propertyKey: string
+): () => void {
+  return function (this: any) {
+    if (!this._memoizeCache) {
+      this._memoizeCache = new Map();
```

Comment:
Is it possible to use `WeakMap` here?

Reply:
No, not without making some other change. WeakMap only accepts objects as keys, but we are using strings as keys for this cache.

![image](https://user-images.githubusercontent.com/2023/117308778-d7235b00-ae4f-11eb-83a4-f545dbf213ea.png)